### ISSUE TITLE
The inheriting keras.layers.Layer does not call a compute_output_shape after switching to tf.keras from keras

### ISSUE THREAD

**Issue by kiflowb777** [tensorflow/tensorflow#33785]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 1.14, 1.15 and 2.0 (gpu)
- Keras version: 2.2.4 and 2.2.4-tf
- Python version: 3.6.1
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10/6.7.4
- GPU model and memory: RTX 2060 6GB

You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` unknown 1.14.0

After switching to tf.keras from keras my project witch use Mask_RCNN does not work.
I noticed that TimeDistributed returns a different tensor shape in keras (1.) and tf.keras (2.):
1.` Tensor("mrcnn_class_bn1/Reshape_1:0", shape=(?, 1000, 1, 1, 256), dtype=float32)`
2. `Tensor("mrcnn_class_bn1/Reshape_1:0", shape=(?, ?, 1, 1, 256), dtype=float32)`

```
x = PyramidROIAlign([pool_size, pool_size], feature_levels,
                        name="roi_align_classifier")([rois, image_meta] + feature_maps)
x = KL.TimeDistributed(KL.Conv2D(fc_layers_size, (pool_size, pool_size), padding="valid"),
                           name="mrcnn_class_conv1")(x)
```

where rois is generated by ProposalLayer from Mask_RCNN module (https://github.com/matterport/Mask_RCNN)

ProposalLayer inherits from the class keras.layers.Layer.
In Keras the overwritten function compute_output_shape is called, but not in the tf.keras.

Additionally, the rois created in the tf.keras does not have the parameter _keras_shape that is most likely needed in the function TimeDistributed to create a proper shape.

**Comment by kiflowb777** [comment]
Related topics:
https://github.com/matterport/Mask_RCNN/issues/1820
https://github.com/keras-team/keras/issues/13497
https://stackoverflow.com/questions/54613474/tensorflow-keras-input-layer-does-not-add-keras-shape

**Comment by ravikyram** [comment]
@kiflowb777 

Can you please share simple standalone code to reproduce the issue in our environment.It helps in localizing the issue faster. Thanks!

**Comment by kiflowb777** [comment]
```
use_tf_keras = False

if use_tf_keras:
    import tensorflow.keras
    from tensorflow.keras.models import Sequential
    from tensorflow.keras import layers
    from tensorflow.keras.datasets import mnist
    from tensorflow.keras import backend as K
else:
    import keras
    from keras.models import Sequential
    from keras import layers
    from keras.datasets import mnist
    from keras import backend as K
    

class Antirectifier(layers.Layer):
    def compute_output_shape(self, input_shape):
        print("compute_output_shape called")
        shape = list(input_shape)
        assert len(shape) == 2  # only valid for 2D tensors
        shape[-1] *= 2
        return tuple(shape)

    def call(self, inputs):
        print("Antirectifier called")
        inputs -= K.mean(inputs, axis=1, keepdims=True)
        inputs = K.l2_normalize(inputs, axis=1)
        pos = K.relu(inputs)
        neg = K.relu(-inputs)
        return K.concatenate([pos, neg], axis=1)

# global parameters
batch_size = 128
num_classes = 10
epochs = 5

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

# build the model
model = Sequential()
model.add(layers.Dense(256, input_shape=(784,)))
model.add(Antirectifier())
model.add(layers.Dropout(0.1))
model.add(layers.Dense(256))
model.add(Antirectifier())
model.add(layers.Dropout(0.1))
model.add(layers.Dense(num_classes))
model.add(layers.Activation('softmax'))

# compile the model
model.compile(loss='categorical_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

# train the model
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
```

If you switch use_tf_keras to True you can notice that compute_output_shape in custom layer is not called.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 1.14 and was able to execute the code  successfully with `use_tf_keras to True ` and `use_tf_keras = False`. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/0c7d4ac2d3a4f820271eec36c8b6dbd7/untitled318.ipynb).I have tried on colab with TF 2.0 and i am seeing the below error.`runtimeError: It looks like you are trying to use a version of multi-backend Keras that does not support TensorFlow 2.0. We recommend using `tf.keras`, or alternatively, downgrading to TensorFlow 1.14.`. Is this is the expected output. Thanks!

**Comment by kiflowb777** [comment]
Thanks.
I'm trying to migrate from keras to tf.keras. Most of my project works on the tf.keras. But one network uses Mask_RCNN (https://github.com/matterport/Mask_RCNN).
I noticed that the layers shapes generated by Mask_RCNN are different if I use the keras and tf.keras.

It seems to me that if you use the tf.keras the compute_output_shape is not called. 
You might notice that in this case that line `print("compute_output_shape called")` is not printed if use tf.keras.

**Comment by kiflowb777** [comment]
You can try run this code in IDE (e.g. PyCharm) and add breakpoint inside `custom compute_output_shape()` during debug mode. 
You may notice that the program will not stop only if you use the tf.keras.

**Comment by kiflowb777** [comment]
I wrote a simple code that shows that the shape differs depending on whether we use the keras or the tf.keras:

```
use_tf_keras = False

if use_tf_keras:
    import tensorflow.keras
    from tensorflow.keras.models import Sequential
    from tensorflow.keras import layers
    from tensorflow.keras import backend as K
else:
    import keras
    from keras.models import Sequential
    from keras import layers
    from keras import backend as K
    
class Antirectifier(layers.Layer):
    def compute_output_shape(self, input_shape):
        print("custom compute_output_shape called")
        shape = list(input_shape)
        shape[-1] *= 2-1
        return tuple(shape)

    def call(self, inputs):
        inputs -= K.mean(inputs, axis=1, keepdims=True)
        inputs = K.l2_normalize(inputs, axis=1)
        pos = K.relu(inputs)
        neg = K.relu(-inputs)
        return K.concatenate([pos, neg], axis=1)

# build the model
model = Sequential()
model.add(layers.Dense(256, input_shape=(784,)))
model.add(Antirectifier())
print(model.output_shape)
```

keras:   **(None, 256)**
tf.keras:   **(None, 512)**

**Comment by kiflowb777** [comment]
I found similar problems:
https://github.com/tensorflow/tensorflow/issues/32029
https://github.com/tensorflow/tensorflow/issues/19961
https://stackoverflow.com/questions/51028861/tensorflow-compute-output-shape-not-working-for-custom-layer

**Comment by kiflowb777** [comment]
I also checked on TF 1,15, but the problem also exists.

**Comment by kiflowb777** [comment]
I tried to manually correct the shape of the custom layer by using the tf.reshape() or set_shape function after each use, but the summary of the model created in tf.keras differs from the one created in keras.

**Comment by qlzh727** [comment]
Sorry for the very late reply. 

I don't think compute_output_shape need to be called if we can already get the output tensor from the layer itself. It doesn't make sense to call compute_output_shape() when we have layer.output available. If the model's output doesn't carry the correct static shape information, I think we should fix the layer.call() body for that.

Could u point me to the layer in your code that produce the unexpected output?

**Comment by qlzh727** [comment]
Ping @kiflowb777 for updates.

**Comment by qlzh727** [comment]
Closing this issue due to lack of feedback, feel free to reopen this issue with more information for further troubleshooting.

**Comment by levyfan** [comment]
Hi @qlzh727 , please reopen this issue and check the following `ResizeLayer` in tensorflow 1.15. It could run on `keras` even if the `compute_output_shape` function is commented out. But in `tf.keras` the code fails as:

```
Traceback (most recent call last):
  File "/Users/fanliwen/PycharmProjects/textual-relevance/layers_test.py", line 58, in <module>
    s = Dense(1)(Dense(20, activation='relu')(z))
  File "/Users/fanliwen/miniconda3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/base_layer.py", line 824, in __call__
    self._maybe_build(inputs)
  File "/Users/fanliwen/miniconda3/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/base_layer.py", line 2146, in _maybe_build
    self.build(input_shapes)
  File "/Users/fanliwen/miniconda3/lib/python3.6/site-packages/tensorflow_core/python/keras/layers/core.py", line 1009, in build
    raise ValueError('The last dimension of the inputs to `Dense` '
ValueError: The last dimension of the inputs to `Dense` should be defined. Found `None`.
```

- use_tf_keras = True: fails
- use_tf_keras = False: works, even `compute_output_shape` not implemented

```
import tensorflow as tf

use_tf_keras = True
if use_tf_keras:
    from tensorflow.keras.layers import Flatten, Dense
    from tensorflow.keras import Input, Model
    from tensorflow.keras import layers
else:
    from keras.layers import Flatten, Dense
    from keras import Input, Model
    from keras import layers


class ResizeLayer(layers.Layer):
    """
    Inputs:
        images:
            4D tensor with shape `(batch_size, rows, cols, channels)`
        rows:
            1D tensor with shape `(batch_size,)`
        cols:
            1D tensor with shape `(batch_size,)`
    Outputs:
        4D tensor with shape `(batch_size, rows, cols, channels)`
    """

    def __init__(self, **kwargs):
        super(ResizeLayer, self).__init__(**kwargs)

    def call(self, inputs, **kwargs):
        images = inputs[0]
        rows = inputs[1]
        cols = inputs[2]

        nhwc = tf.shape(images)
        boxes = tf.stack([tf.zeros(shape=(nhwc[0],), dtype='float32'),
                          tf.zeros(shape=(nhwc[0],), dtype='float32'),
                          tf.cast(rows, 'float32') / tf.cast(nhwc[1], 'float32'),
                          tf.cast(cols, 'float32') / tf.cast(nhwc[2], 'float32'), ], axis=-1)
        box_ind = tf.range(0, limit=nhwc[0])
        crop_size = (nhwc[1], nhwc[2])

        resize = tf.image.crop_and_resize(images, boxes, box_ind, crop_size)

        return resize

    def compute_output_shape(self, input_shape):
        return input_shape


img = Input(shape=(8, 4, 3), name='img')
q_len = Input(batch_shape=(None,), name='q_len')
d_len = Input(batch_shape=(None,), name='d_len')

z2 = ResizeLayer()([img, q_len, d_len])

z = Flatten()(z2)
s = Dense(1)(Dense(20, activation='relu')(z))
model = Model([img, q_len, d_len], s, name='resize')
print(model.summary())
```

**Comment by qlzh727** [comment]
@levyfan, the issue is caused by the line:
crop_size = (nhwc[1], nhwc[2])

Note that nhwc is a tensor and its slice will not have the value when feed as crop_size, and it cause the resize shape to be [None, None, None, 3], rather than [None, 8, 4, 3].

If you replace the line with below, which use image's static size, then it will correctly propagate the static shape to the resize.
crop_size = images.shape[1:3]

Alternatively, you can also just set the static shape of resize before u return:
resize.shape = images.shape
return resize

**Comment by levyfan** [comment]
Thanks, that works!

**Comment by kartik4949** [comment]
@qlzh727  Hi,
above problems about matterport maskrcnn is stated below


if we use keras - 2.2.4
the output produced by timedistributed layer is
x - inputshape --> (1,None,7,7,256)
x = keras.Timedistributed(conv2(1024 , (7,7))) (x)
x - output shape --> (None , 200, 1,1,1024)

for tensorflow 2.0
the output produced by timedistributed layer is
x - inputshape --> (1,None,7,7,256)
x = tf.keras.Timedistributed(conv2(1024 , (7,7))) (x)
x - output shape --> (None , None, 1,1,1024)

so basically the output shape is wrong in new tf 2.0 /timesteps calculated wrongly

any ideas how to solvethis
thanks

**Comment by kartik4949** [comment]
> I tried to manually correct the shape of the custom layer by using the tf.reshape() or set_shape function after each use, but the summary of the model created in tf.keras differs from the one created in keras.

hi i have converted the matterport model to tf 2.x model
just had few queries , will you help me
Thanks

**Comment by Lucysmith8** [comment]
Do not help this person. When he finds the answer, he closes the question and does not respond to those who need help.

**Comment by kiflowb777** [comment]
I found a solution, try at the end of the self.call method:

```
def call(self, input):
   input_shape = input.shape  # you must find a way to retrieve this value on your own
   ...
   out_shape = self.compute_output_shape(input_shape)
   output = output.set_shape(out_shape)
   return output
```
